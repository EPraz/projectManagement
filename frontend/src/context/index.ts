export * from "./sprintContext";
export * from "./apiContext";
export * from "./projectContext";
export * from './snackbarContext'