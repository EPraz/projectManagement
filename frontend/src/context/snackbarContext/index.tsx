import { createContext, useContext, useState, ReactNode } from "react";
import { Snackbar, Alert } from "@mui/material";

type SnackbarContextType = {
  showSnackbarMessage: (
    message: string,
    type?: "success" | "error" | "info"
  ) => void;
};

const SnackbarContext = createContext<SnackbarContextType | undefined>(
  undefined
);

export const SnackbarProvider = ({ children }: { children: ReactNode }) => {
  const [snackbar, setSnackbar] = useState<{
    open: boolean;
    message: string;
    type: "success" | "error" | "info";
  }>({ open: false, message: "", type: "info" });

  const showSnackbarMessage = (
    message: string,
    type: "success" | "error" | "info" = "info"
  ) => {
    setSnackbar({ open: true, message, type });
  };

  return (
    <SnackbarContext.Provider value={{ showSnackbarMessage }}>
      {children}
      <Snackbar
        open={snackbar.open}
        autoHideDuration={2500}
        onClose={() => setSnackbar((prev) => ({ ...prev, open: false }))}
        anchorOrigin={{ vertical: "top", horizontal: "right" }}
      >
        <Alert
          severity={snackbar.type}
          onClose={() => setSnackbar((prev) => ({ ...prev, open: false }))}
        >
          {snackbar.message}
        </Alert>
      </Snackbar>
    </SnackbarContext.Provider>
  );
};

export const useSnackbar = () => {
  const context = useContext(SnackbarContext);
  if (!context)
    throw new Error("useSnackbar must be used within a SnackbarProvider");
  return context;
};
