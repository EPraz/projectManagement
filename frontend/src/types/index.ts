export * from "./contextProps";
export * from "./models";
export * from "./materialUiExtensions";
export * from "./components";
export * from "./FilterProps";
