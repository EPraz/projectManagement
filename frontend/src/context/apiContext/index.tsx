import { createContext, useContext } from "react";
import { API_URL } from "../../constants";
import { ApiContextType } from "../../types";

const ApiContext = createContext<ApiContextType>({ apiUrl: API_URL });

export const useApi = () => useContext(ApiContext);

export const ApiProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ApiContext.Provider value={{ apiUrl: API_URL }}>
      {children}
    </ApiContext.Provider>
  );
};
