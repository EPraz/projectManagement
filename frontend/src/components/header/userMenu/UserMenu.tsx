import React, { useState } from "react";
import {
  Avatar,
  Box,
  IconButton,
  Menu,
  MenuItem,
  Divider,
  Typography,
  ListItemIcon,
} from "@mui/material";
import {
  Logout as LogoutIcon,
  Settings as SettingsIcon,
  Person as PersonIcon,
} from "@mui/icons-material";
import { useNavigate } from "react-router-dom";
import { useAuth } from "../../../context";
import { getInitials } from "../../../helpers";

const UserMenu = () => {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const { user, logout } = useAuth();
  const navigate = useNavigate();

  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const handleLogout = () => {
    logout();
    handleClose();
    navigate("/auth/login");
  };

  const handleProfile = () => {
    // navigate("/profile");
    handleClose();
  };

  const handleSettings = () => {
    // navigate("/settings");
    handleClose();
  };

  return (
    <>
      <IconButton
        onClick={handleClick}
        size="small"
        aria-controls={open ? "user-menu" : undefined}
        aria-haspopup="true"
        aria-expanded={open ? "true" : undefined}
        sx={{
          ml: 2,
          border: "2px solid",
          borderColor: "primary.main",
          transition: "all 0.2s",
          width: "fit-content",
          "&:hover": {
            backgroundColor: "primary.main",
            color: "white",
          },
        }}
      >
        <Avatar
          sx={{
            width: 32,
            height: 32,
            bgcolor: "primary.main",
            color: "white",
            fontSize: "0.875rem",
            fontWeight: "bold",
          }}
        >
          {getInitials(user?.name)}
        </Avatar>
      </IconButton>
      <Menu
        anchorEl={anchorEl}
        id="user-menu"
        open={open}
        onClose={handleClose}
        // onClick={handleClose}
        PaperProps={{
          elevation: 3,
          sx: {
            overflow: "visible",
            filter: "drop-shadow(0px 2px 8px rgba(0,0,0,0.15))",
            mt: 1.5,
            width: 220,
            borderRadius: 2,
            "& .MuiAvatar-root": {
              width: 32,
              height: 32,
              ml: -0.5,
              mr: 1,
            },
          },
        }}
        transformOrigin={{ horizontal: "right", vertical: "top" }}
        anchorOrigin={{ horizontal: "right", vertical: "bottom" }}
      >
        <Box sx={{ px: 2, py: 1.5 }}>
          <Typography variant="subtitle1" fontWeight="bold">
            {user?.name || "User"}
          </Typography>
          <Typography variant="body2" color="text.secondary">
            {user?.email || "user@example.com"}
          </Typography>
        </Box>
        <Divider />
        <MenuItem onClick={handleProfile} disabled={true}>
          <ListItemIcon>
            <PersonIcon fontSize="small" />
          </ListItemIcon>
          Profile
        </MenuItem>
        <MenuItem onClick={handleSettings} disabled={true}>
          <ListItemIcon>
            <SettingsIcon fontSize="small" />
          </ListItemIcon>
          Settings
        </MenuItem>
        <Divider />
        <MenuItem onClick={handleLogout}>
          <ListItemIcon>
            <LogoutIcon fontSize="small" color="error" />
          </ListItemIcon>
          <Typography color="error.main">Logout</Typography>
        </MenuItem>
      </Menu>
    </>
  );
};

export default UserMenu;
