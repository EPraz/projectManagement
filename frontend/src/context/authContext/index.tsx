import React, {
  createContext,
  useContext,
  useState,
  useEffect,
  ReactNode,
  useMemo,
} from "react";
import { useLocation, useNavigate } from "react-router-dom";
import { useApi } from "../apiContext";
import { User } from "../../types";

interface AuthContextType {
  user: User | null;
  login: (email: string, password: string) => Promise<void>;
  register: (email: string, password: string, name: string) => Promise<void>;
  logout: () => Promise<void>;
  instantLogin: (email: string) => Promise<void>;
  loading: boolean;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export const AuthProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const { apiUrl } = useApi();
  const [user, setUser] = useState<User | null>(null);
  const [loading, setLoading] = useState<boolean>(true);
  const [accessToken, setAccessToken] = useState<string | null>(null);
  const [initialized, setInitialized] = useState<boolean>(false);

  const navigate = useNavigate();
  const location = useLocation();
  const isLoginPage = useMemo(
    () => /^\/auth\/login/.test(location.pathname),
    [location.pathname]
  );

  useEffect(() => {
    const refreshTokenIfNeeded = async () => {
      try {
        const refreshResponse = await fetch(`${apiUrl}/auth/refresh`, {
          method: "POST",
          credentials: "include", // Envia la cookie con el refresh token
          headers: { "Content-Type": "application/json" },
        });
        if (refreshResponse.ok) {
          const data: { accessToken: string; refreshToken: string } =
            await refreshResponse.json();
          setAccessToken(data.accessToken);
        }
      } catch (error) {
        console.error("Error en refresh:", error);
      } finally {
        setInitialized(true);
      }
    };
    if (!isLoginPage) {
      refreshTokenIfNeeded();
    } else {
      setInitialized(true);
    }
  }, [apiUrl]);

  useEffect(() => {
    const checkAuth = async () => {
      if (!accessToken) {
        setLoading(false);
        setUser(null);
        return;
      }
      setLoading(true);
      try {
        const response = await fetch(`${apiUrl}/auth/checkAuth`, {
          method: "GET",
          credentials: "include", // Para enviar cookies
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${accessToken}`,
          },
        });
        if (response.ok) {
          const data: User = await response.json();
          setUser(data);
        } else {
          setUser(null);
        }
      } catch (error) {
        console.error("Error en checkAuth:", error);
        setUser(null);
      } finally {
        setLoading(false);
      }
    };

    if (initialized) {
      checkAuth();
    }
  }, [apiUrl, accessToken, initialized]);

  const login = async (email: string, password: string) => {
    setLoading(true);

    const response = await fetch(`${apiUrl}/auth/login`, {
      method: "POST",
      credentials: "include",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email, password }),
    });

    setLoading(false);
    if (!response.ok) throw new Error("Invalid credentials");

    const { accessToken: token } = await response.json();
    setAccessToken(token);
    navigate("/dashboard");
  };

  const register = async (email: string, password: string, name: string) => {
    setLoading(true);

    const response = await fetch(`${apiUrl}/auth/register`, {
      method: "POST",
      credentials: "include",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email, password, name }),
    });

    setLoading(false);
    if (!response.ok) throw new Error("Invalid credentials");

    const { accessToken: token } = await response.json();
    setAccessToken(token);
    navigate("/dashboard");
  };

  const instantLogin = async (email: string) => {
    setLoading(true);

    const response = await fetch(`${apiUrl}/auth/instant-login`, {
      method: "POST",
      credentials: "include",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email }),
    });

    setLoading(false);
    if (!response.ok) throw new Error("Failed to login instantly");

    const { accessToken: token } = await response.json();
    setAccessToken(token);
    navigate("/dashboard");
  };

  const logout = async () => {
    await fetch(`${apiUrl}/auth/logout`, {
      method: "POST",
      credentials: "include",
      headers: { "Content-Type": "application/json" },
    });

    setUser(null);
    setAccessToken(null);
    setLoading(false);
    navigate("/auth/login");
  };

  return (
    <AuthContext.Provider
      value={{ user, login, logout, instantLogin, register, loading }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) throw new Error("useAuth must be used within an AuthProvider");
  return context;
};
