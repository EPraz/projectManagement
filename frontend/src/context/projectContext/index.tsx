import {
  createContext,
  useContext,
  useState,
  ReactNode,
  useEffect,
  useCallback,
  useMemo,
} from "react";
import { useParams } from "react-router-dom";
import { Project, Sprint, Ticket } from "../../types";
import { useApi } from "../apiContext";
import { Loading } from "../../components";

const ProjectContext = createContext<{
  project: Project | null;
  loading: boolean;
  listOfSprints: Sprint[];
  setListOfSprints: React.Dispatch<React.SetStateAction<Sprint[]>>;
  setProject: React.Dispatch<React.SetStateAction<Project | null>>;
} | null>(null);

export const ProjectProvider = ({ children }: { children: ReactNode }) => {
  const { id } = useParams();
  const { apiUrl } = useApi();

  const [project, setProject] = useState<Project | null>(null);
  const [loading, setLoading] = useState(false);
  const [listOfSprints, setListOfSprints] = useState<Sprint[]>([]);

  //  Verificar si estamos en una pÃ¡gina de proyecto antes de cargar datos
  const isProjectPage = useMemo(
    () => /^\/projects\/[^/]+/.test(window.location.pathname),
    [window.location.pathname]
  );

  const loadProject = useCallback(async () => {
    if (!id || !isProjectPage) return;
    setLoading(true);
    try {
      const response = await fetch(`${apiUrl}/projects/${id}`, {
        method: "GET",
        headers: { "Content-Type": "application/json" },
      });
      if (!response.ok) throw new Error("Failed to fetch project");
      const data: Project = await response.json();
      setProject(data);
      setListOfSprints(data.sprints);
    } catch (error) {
      console.error("Error loading project:", error);
    } finally {
      setLoading(false);
    }
  }, [id, apiUrl, isProjectPage]);

  useEffect(() => {
    loadProject();
  }, [loadProject]);

  useEffect(() => {
    setProject((prev) => {
      if (!prev) return prev;
      return { ...prev, sprints: listOfSprints };
    });
  }, [listOfSprints]);

  const memoizedProject = useMemo(() => project, [project]);

  if (loading) return <Loading message="Cargando proyecto..." />;

  return (
    <ProjectContext.Provider
      value={{
        project: memoizedProject,
        loading,
        listOfSprints,
        setListOfSprints,
        setProject,
      }}
    >
      {children}
    </ProjectContext.Provider>
  );
};

export const useProject = () => {
  const context = useContext(ProjectContext);
  if (!context)
    throw new Error("useProject must be used within ProjectProvider");
  return context;
};
