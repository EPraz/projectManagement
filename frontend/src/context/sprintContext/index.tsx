import {
  createContext,
  useContext,
  useState,
  ReactNode,
  useEffect,
  useCallback,
} from "react";
import { Sprint, SprintContextProps, Ticket } from "../../types";
import { useApi } from "../apiContext";
import { useProject } from "../projectContext";
import { Loading } from "../../components";

const SprintContext = createContext<SprintContextProps | undefined>(undefined);

export const SprintProvider = ({ children }: { children: ReactNode }) => {
  const { apiUrl } = useApi();
  const { project, loading: projectLoading } = useProject(); // Obtenemos `project` y `loading`
  const [sprint, setSprint] = useState<Sprint | null>(null);
  const [tickets, setTickets] = useState<Ticket[]>([]);
  const [loading, setLoading] = useState(true); // Loading para el Sprint

  useEffect(() => {
    if (projectLoading) return; // ðŸ”¹ Esperamos a que `project` termine de cargar

    if (!project || project.sprints.length === 0) {
      setSprint(null);
      setLoading(false);
      return;
    }

    const today = new Date();
    const activeSprint = project.sprints.find(
      (s) => new Date(s.startDate) <= today && new Date(s.endDate) >= today
    );

    setSprint(activeSprint || project.sprints[0]); // Si no hay Sprint activo, usa el primero
    setLoading(false);
  }, [project, projectLoading]);

  const loadTickets = useCallback(async () => {
    if (!sprint?.id) return;
    setLoading(true);
    try {
      const response = await fetch(`${apiUrl}/tickets?sprintId=${sprint.id}`);
      if (!response.ok) throw new Error("Failed to fetch tickets");
      const data = await response.json();
      setTickets(data);
    } catch (error) {
      console.error("Error loading tickets:", error);
    } finally {
      setLoading(false);
    }
  }, [sprint, apiUrl]);

  useEffect(() => {
    loadTickets();
  }, [loadTickets]);

  if (loading) return <Loading message="Cargando sprint y tickets..." />;

  return (
    <SprintContext.Provider value={{ sprint, setSprint, tickets, loadTickets }}>
      {children}
    </SprintContext.Provider>
  );
};

export const useSprint = () => {
  const context = useContext(SprintContext);
  if (!context)
    throw new Error("useSprint must be used within a SprintProvider");
  return context;
};
