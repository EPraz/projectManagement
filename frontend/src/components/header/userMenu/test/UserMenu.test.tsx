// UserMenu.test.tsx
import { render, screen, fireEvent } from "@testing-library/react";
import UserMenu from "../UserMenu";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "../../../../context";
import { User } from "../../../../types";

// Mockea useNavigate de react-router-dom
const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => {
  const actual = jest.requireActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});

// Mockea getInitials para simplificar la prueba
jest.mock("../../../../helpers", () => ({
  getInitials: (name: string | undefined) => (name ? name[0] : "U"),
}));

describe("UserMenu Component", () => {
  const mockLogout = jest.fn();

  // Define un usuario de prueba
  const user: User = {
    id: "1",
    name: "John Doe",
    email: "john@example.com",
    projects: [],
  };

  // Crea una función para renderizar el componente envuelto en el AuthContext y MemoryRouter
  const renderWithAuth = () =>
    render(
      <AuthContext.Provider
        value={{
          user,
          logout: mockLogout,
          login: jest.fn(),
          register: jest.fn(),
          instantLogin: jest.fn(),
          loading: false,
          accessToken: "fake-token",
        }}
      >
        <MemoryRouter>
          <UserMenu />
        </MemoryRouter>
      </AuthContext.Provider>
    );

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("match snapshot", async () => {
    const { asFragment } = renderWithAuth();
    expect(asFragment).toMatchSnapshot();
  });

  it("renders the user avatar with initials", () => {
    renderWithAuth();
    // El avatar debería mostrar la inicial "J" (de John)
    expect(screen.getByText("J")).toBeInTheDocument();
  });

  it("opens menu when avatar is clicked", () => {
    renderWithAuth();
    const avatarButton = screen.getByRole("button");
    fireEvent.click(avatarButton);
    // Se espera que aparezca el menú, que tiene role "menu"
    expect(screen.getByRole("menu")).toBeInTheDocument();
  });

  it("calls logout and navigates on clicking Logout", () => {
    renderWithAuth();
    // Abrir el menú
    const avatarButton = screen.getByRole("button");
    fireEvent.click(avatarButton);
    // Buscar el menú y luego la opción de Logout (buscamos por el texto "Logout")
    const logoutMenuItem = screen.getByText("Logout");
    fireEvent.click(logoutMenuItem);
    expect(mockLogout).toHaveBeenCalled();
    expect(mockNavigate).toHaveBeenCalledWith("/auth/login");
  });

  it("renders disabled Profile and Settings menu items", () => {
    renderWithAuth();
    fireEvent.click(screen.getByRole("button"));
    // Los MenuItem con texto "Profile" y "Settings" están deshabilitados.
    const profileMenuItem = screen.getByText("Profile").closest("li");
    const settingsMenuItem = screen.getByText("Settings").closest("li");
    // Material UI agrega la propiedad disabled como un atributo en el DOM (a veces 'aria-disabled')
    expect(profileMenuItem).toHaveAttribute("aria-disabled", "true");
    expect(settingsMenuItem).toHaveAttribute("aria-disabled", "true");
  });
});
