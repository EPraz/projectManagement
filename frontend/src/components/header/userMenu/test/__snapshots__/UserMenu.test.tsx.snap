// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMenu Component match snapshot 1`] = `[Function]`;
